# Import Libraries

In [3]:
import pandas as pd
import numpy as np
import os

# Importing Data

In [6]:
path = r'C:\Users\shaiv\Downloads\Instacart Basket Analysis'

In [8]:
path

'C:\\Users\\shaiv\\Downloads\\Instacart Basket Analysis'

In [10]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [12]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Data Consistency Checks

# Mixed-Type Data

In [23]:
# Creating a mix data type column in a test dataframe
df_test = pd.DataFrame()

In [25]:
# The mixed column is filled with different types of data
df_test['mix'] = ['a','b', 1, True]

In [29]:
# Function to search for a mixed data type in a column
for col in df_test.columns.tolist():
    weird = (df_test[[col]]. applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\shaiv\AppData\Local\Temp\ipykernel_9672\1715989080.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]]. applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [31]:
# Converting mixed column data type to String
df_test['mix'] = df_test['mix'].astype('str')

In [33]:
df_test['mix'].dtype

dtype('O')

# Find Missing Values in Products Data Set

In [35]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [37]:
#Creating a datafarme to subset the data in the df_prods where there  missing values are found  in the product_name column
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [39]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [41]:
# Running explanatory analysis on the df_products dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [43]:
# Median of df_prods dataframe
df_prods['prices'].median()

7.1

In [47]:
# Finding outliers in the df_prods dataframe
df_prods[df_prods['prices'] > df_prods['prices'].mean() + (2 * df_prods['prices'].std())]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [49]:
# Replacing outliers values with the median value
df_prods = df_prods.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [51]:
# Running explanatory analysis on the df_products dataframe
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682245
std    14343.717401     38.316774       5.850282      4.200159
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

# Observation: Exploratory Analysis on Products DataFrame
An exploratory analysis was conducted on the df_products DataFrame, revealing the presence of outliers in the data. The prices column had a mean of 9.99 and a standard deviation of 453.6. The high standard deviation, which is more than two standard deviations from the mean, indicated the existence of outlier values.

Identifying all data points in the prices column that were two standard deviations away from the mean highlighted two outliers: prices at 14,900 and 99,999. These outliers were replaced with the median value of 7.1 to mitigate their impact on the data distribution.

Given the wide range of values, from a minimum of 1 to a maximum of 99,999, replacing outliers with the median is recommended. The median is less sensitive to extreme values compared to the mean, making it a more robust measure in this context.

After replacing the outliers with the median, the mean of the prices column decreased to 7.6, reflecting a more balanced data set.

In [55]:
#Checking the number of rows & columns in df_prods dataframe
df_prods.shape

(49693, 5)

In [57]:
#Creating a new dataframe as a subset of df_prods dataframe where the missing values in the product_name variable are excluded
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [59]:
df_prods_clean.shape

(49677, 5)

# Removing Duplicate Values

In [62]:
# Finding duplicate values
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [64]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [66]:
# Number of rows in the df_prods_clean dataframe
df_prods_clean.shape

(49677, 5)

In [68]:
# New dataframe with no duplicate values
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [70]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 5

In [78]:
# 02 Running explanatory analysis on the df_products dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Upon reviewing the df_ords DataFrame, I meticulously examined the values in each column to ensure their validity. No unreasonable or outlier values were detected. All entries in the columns are within an acceptable range and are interpretable, corresponding appropriately to their respective variables.

# 03 Mixed-type Data in df_ords

In [85]:
# Fucntion to search for a mixed data type in a column in the df_ords dataframe
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
if len (df_ords[weird]) > 0:
    print (col)

C:\Users\shaiv\AppData\Local\Temp\ipykernel_9672\3100317283.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\shaiv\AppData\Local\Temp\ipykernel_9672\3100317283.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\shaiv\AppData\Local\Temp\ipykernel_9672\3100317283.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\shaiv\AppData\Local\Temp\ipykernel_9672\3100317283.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\shaiv\AppData\Local\Tem

In [87]:
#Checking the data types of all the columns in the df_ords dataframe
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


# Observation
An analysis of the df_ords DataFrame confirmed the absence of mixed-type data in any of its columns or variables. This was validated using a function specifically designed to detect mixed-type data, which returned NULL. Additionally, the info() function further confirmed that all columns maintain consistent data types. No mixed-type data was detected in the DataFrame.

# 05 Missing values in df_ords dataframe

In [91]:
# Check missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [93]:
df_ords

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
1                 1   2398795        1             2          3   
2                 2    473747        1             3          3   
3                 3   2254736        1             4          4   
4                 4    431534        1             5          4   
...             ...       ...      ...           ...        ...   
3421078     3421078   2266710   206209            10          5   
3421079     3421079   1854736   206209            11          4   
3421080     3421080    626363   206209            12          1   
3421081     3421081   2977660   206209            13          1   
3421082     3421082    272231   206209            14          6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [95]:
#Creating a datafarme to subset the data in the df_ords where there  missing values are found in the days_since_prior_order column
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [97]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
11               11   2168274        2             1          2   
26               26   1374495        3             1          1   
39               39   3343014        4             1          6   
45               45   2717275        5             1          3   
...             ...       ...      ...           ...        ...   
3420930     3420930    969311   206205             1          4   
3420934     3420934   3189322   206206             1          3   
3421002     3421002   2166133   206207             1          6   
3421019     3421019   2227043   206208             1          1   
3421069     3421069   3154581   206209             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

# Now we must verify whether all 'order_number' values equal to 1 are missing values in the 'days_since_prior_order' column

In [102]:
# Filter the dataframe where 'order_number' equals 1
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [106]:
# Check if all of these have missing values in 'days_since_prior_order'
df_first_orders_missing = df_first_orders['days_since_prior_order'].isna().all()

In [108]:
# Result
print("All first orders have missing 'days_since_prior_order':", df_first_orders_missing)

All first orders have missing 'days_since_prior_order': True


# Every row with 'order_number' equal to 1 has a missing value

In [114]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [116]:
df_ords

Unnamed: 0  order_id  user_id  order_number  order_dow  \
0                 0   2539329        1             1          2   
1                 1   2398795        1             2          3   
2                 2    473747        1             3          3   
3                 3   2254736        1             4          4   
4                 4    431534        1             5          4   
...             ...       ...      ...           ...        ...   
3421078     3421078   2266710   206209            10          5   
3421079     3421079   1854736   206209            11          4   
3421080     3421080    626363   206209            12          1   
3421081     3421081   2977660   206209            13          1   
3421082     3421082    272231   206209            14          6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 8 columns]

In [118]:
# Check data-type of new column and update if necessary
df_ords['first_order'].dtype

dtype('bool')

In [120]:
# Re-check data to make sure totals are accurate
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

# Missing values in the order_number column have been appropriately addressed, allowing the client to understand their significance. Removing or filtering out these values would have been incorrect, as the missing data holds valuable insights.

# 07 Looking for duplicates

In [128]:
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order, first_order]
Index: []

# 08 There are no duplicate values found in the df_ords dataframe.

In [131]:
# Removing column "Unnamed: 0" from the df_ords dataframe and assigning a new dataframe
df_ords_clean = df_ords.drop('Unnamed: 0', axis = 1)

In [133]:
df_ords_clean

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  first_order  
0                           NaN         True  
1                          15.0        False  
2                          21.0        False  
3                          29.0        False  
4                          28.0        False  
...                         ...          ...  
3421078                    29.0        False  
3421079                    30.0        False  
3421080                    18.0        False  
3421081                     7.0        False  
3421082                    30.0        False  

[3421083 rows x 7 columns]

# 09 Exporting data set

In [136]:
# exporting product data set
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [138]:
# exporting order data set
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))